python/ray/experimental/raysort/main.py

```diff
@@ -190,146 +212,246 @@ def _dummy_sort_and_partition(part: np.ndarray,
 
 
 @ray.remote
-def mapper(boundaries: List[int], mapper_id: PartId,
-           path: Path) -> List[ray.ObjectRef]:
+@tracing_utils.timeit("map")
+def mapper(args: Args, mapper_id: PartId, boundaries: List[int],
+           path: Path) -> List[np.ndarray]:
     logging_utils.init()
-    task_id = f"M-{mapper_id} Mapper"
-    logging.info(f"{task_id} starting {args}")
-    if args.skip_input:
-        block_size = int(np.ceil(args.input_part_size / args.num_reducers))
-        return [
-            ray.put(
-                np.frombuffer(np.random.bytes(block_size), dtype=np.uint8))
-            for _ in range(args.num_reducers)
-        ]
-
-    part = _load_partition(path)
+    part = _load_partition(args, path)
     sort_fn = _dummy_sort_and_partition \
         if args.skip_sorting else sortlib.sort_and_partition
     blocks = sort_fn(part, boundaries)
-    logging.info(f"{task_id} saving to object store")
-    return [ray.put(part[offset:offset + size]) for offset, size in blocks]
+    return [part[offset:offset + size] for offset, size in blocks]
 
 
-def _dummy_merge(blocks: List[np.ndarray], _n: int) -> Iterable[memoryview]:
-    for block in blocks:
+def _dummy_merge(
+        num_blocks: int, _n: int,
+        get_block: Callable[[int, int], np.ndarray]) -> Iterable[np.ndarray]:
+    blocks = [((i, 0), get_block(i, 0)) for i in range(num_blocks)]
+    while len(blocks) > 0:
+        (m, d), block = blocks.pop(random.randrange(len(blocks)))
         yield block
-
-
-@ray.remote
-def reducer(reducer_id: PartId, *blocks: List[ray.ObjectRef]) -> PartitionInfo:
-    logging_utils.init()
-    task_id = f"R-{reducer_id} Reducer"
-    logging.info(f"{task_id} starting")
-    blocks = [np.copy(ray.get(block)) for block in blocks]
+        d_ = d + 1
+        block = get_block(m, d_)
+        if block is None:
+            continue
+        blocks.append(((m, d_), block))
+
+
+def _merge_impl(args: Args,
+                M: int,
+                pinfo: PartInfo,
+                get_block: Callable[[int, int], np.ndarray],
+                skip_output=False):
     merge_fn = _dummy_merge if args.skip_sorting else sortlib.merge_partitions
-    merger = merge_fn(blocks, args.reducer_batch_num_records)
-    if args.skip_output:
+    merger = merge_fn(M, args.reducer_output_chunk, get_block)
+
+    if skip_output:
         for datachunk in merger:
             del datachunk
-        logging.info(f"{task_id} done")
-        return None
     else:
-        pinfo = _make_partition_info(reducer_id, "output")
         with open(pinfo.path, "wb") as fout:
             for datachunk in merger:
                 fout.write(datachunk)
-        logging.info(f"{task_id} done")
-        return pinfo
+    return pinfo
 
 
-@tracing_utils.timeit("sorting")
-def sort_main():
-    partitions = _load_manifest(constants.INPUT_MANIFEST_FILE)
+@ray.remote(
+    num_cpus=0, resources={"worker": 1e-3})  # TODO: test setting this to 1
```

Comment:
How about setting "worker" to 1 now?

Can you also add a note to explain why we use num_cpus=0?